File: test/evm/stack_test.zig
const std = @import("std");
const testing = std.testing;
const evm = @import("evm");
const Stack = evm.Stack;

fn setup_stack(allocator: std.mem.Allocator, items: []const u256) !Stack {
    var stack = try Stack.init(allocator);
    for (items) |item| {
        try stack.append(item);
    }
    return stack;
}

test "Stack: initialization" {
    var stack = try Stack.init(testing.allocator);
    defer stack.deinit(std.testing.allocator);
    try testing.expectEqual(@as(usize, 0), stack.size());
    try testing.expectEqual(@as(usize, Stack.CAPACITY), Stack.CAPACITY);
}

test "Stack: basic push and pop operations" {
    var stack = try Stack.init(testing.allocator);
    defer stack.deinit(std.testing.allocator);

    // Push values
    try stack.append(1);
    try stack.append(2);
    try stack.append(3);

    // Check state
    try testing.expectEqual(@as(usize, 3), stack.size());

    // Pop values and verify LIFO order
    try testing.expectEqual(@as(u256, 3), try stack.pop());
    try testing.expectEqual(@as(u256, 2), try stack.pop());
    try testing.expectEqual(@as(u256, 1), try stack.pop());

    // Stack should be empty
    try testing.expectEqual(@as(usize, 0), stack.size());

    // Pop from empty stack should error
    try testing.expectError(Stack.Error.StackUnderflow, stack.pop());
}

test "Stack: push and pop (safe variants)" {
    var stack = try Stack.init(testing.allocator);
    defer stack.deinit(std.testing.allocator);

    // Safe operations first to set up state
    try stack.append(100);
    try stack.append(200);

    // Push using safe API
    try stack.append(300);
    try testing.expectEqual(@as(usize, 3), stack.size());

    const value = try stack.pop();
    try testing.expectEqual(@as(u256, 300), value);
    try testing.expectEqual(@as(usize, 2), stack.size());
}

test "Stack: peek operations" {
    var stack = try setup_stack(testing.allocator, &[_]u256{ 1, 2, 3, 4, 5 });
    defer stack.deinit(std.testing.allocator);

    // Test peek (top element)
    const top = try stack.peek();
    try testing.expectEqual(@as(u256, 5), top);
    try testing.expectEqual(@as(usize, 5), stack.size()); // Size unchanged

    // Test peek_n (nth element from top)
    try testing.expectEqual(@as(u256, 5), try stack.peek_n(0)); // Top
    try testing.expectEqual(@as(u256, 4), try stack.peek_n(1)); // Second from top
    try testing.expectEqual(@as(u256, 1), try stack.peek_n(4)); // Bottom

    // Out of bounds should error
    try testing.expectError(Stack.Error.StackUnderflow, stack.peek_n(5));
}

test "Stack: dup operation" {
    var stack = try setup_stack(testing.allocator, &[_]u256{ 10, 20, 30 });
    defer stack.deinit(std.testing.allocator);

    // Duplicate top element (n=1)
    try stack.dup(1);
    try testing.expectEqual(@as(usize, 4), stack.size());
    try testing.expectEqual(@as(u256, 30), try stack.pop()); // Duplicated top
    try testing.expectEqual(@as(u256, 30), try stack.pop()); // Original top

    // Test duplicating deeper element
    try stack.dup(2); // Should duplicate 10 (2nd from top)
    try testing.expectEqual(@as(u256, 10), try stack.pop());
}

test "Stack: swap operation" {
    var stack = try setup_stack(testing.allocator, &[_]u256{ 1, 2, 3, 4, 5 });
    defer stack.deinit(std.testing.allocator);

    // Swap top with second element (n=1)
    try stack.swap(1);
    try testing.expectEqual(@as(u256, 4), try stack.pop()); // Was second, now top
    try testing.expectEqual(@as(u256, 5), try stack.pop()); // Was top, now second

    // Verify remaining elements unchanged
    try testing.expectEqual(@as(u256, 3), try stack.pop());
    try testing.expectEqual(@as(u256, 2), try stack.pop());
    try testing.expectEqual(@as(u256, 1), try stack.pop());
}

test "Stack: pop2 operation" {
    var stack = try setup_stack(testing.allocator, &[_]u256{ 1, 2, 3, 4, 5 });
    defer stack.deinit(std.testing.allocator);

    const popped = try stack.pop2();
    try testing.expectEqual(@as(u256, 4), popped.a); // Second from top
    try testing.expectEqual(@as(u256, 5), popped.b); // Top
    try testing.expectEqual(@as(usize, 3), stack.size());

    // Verify remaining elements
    try testing.expectEqual(@as(u256, 3), try stack.pop());
}

test "Stack: pop3 operation" {
    var stack = try setup_stack(testing.allocator, &[_]u256{ 1, 2, 3, 4, 5 });
    defer stack.deinit(std.testing.allocator);

    const popped = try stack.pop3();
    try testing.expectEqual(@as(u256, 3), popped.a); // Third from top
    try testing.expectEqual(@as(u256, 4), popped.b); // Second from top
    try testing.expectEqual(@as(u256, 5), popped.c); // Top
    try testing.expectEqual(@as(usize, 2), stack.size());

    // Verify remaining elements
    try testing.expectEqual(@as(u256, 2), try stack.pop());
    try testing.expectEqual(@as(u256, 1), try stack.pop());
}

test "Stack: set_top operation" {
    var stack = try setup_stack(testing.allocator, &[_]u256{ 1, 2, 3 });
    defer stack.deinit(std.testing.allocator);

    try stack.set_top(999);
    try testing.expectEqual(@as(u256, 999), try stack.pop());
    try testing.expectEqual(@as(u256, 2), try stack.pop());
    try testing.expectEqual(@as(u256, 1), try stack.pop());
}

test "Stack: clear operation" {
    var stack = try setup_stack(testing.allocator, &[_]u256{ 1, 2, 3, 4, 5 });
    defer stack.deinit(std.testing.allocator);
    try testing.expectEqual(@as(usize, 5), stack.size());

    stack.clear();
    try testing.expectEqual(@as(usize, 0), stack.size());

    // Should be able to use stack normally after clear
    try stack.append(100);
    try testing.expectEqual(@as(usize, 1), stack.size());
    try testing.expectEqual(@as(u256, 100), try stack.pop());
}

test "Stack: overflow protection" {
    var stack = try Stack.init(testing.allocator);
    defer stack.deinit(std.testing.allocator);

    // Fill stack to capacity - 1
    for (0..Stack.CAPACITY - 1) |i| {
        try stack.append(@intCast(i));
    }
    try testing.expectEqual(@as(usize, Stack.CAPACITY - 1), stack.size());

    // This should succeed
    try stack.append(999);
    try testing.expectEqual(@as(usize, Stack.CAPACITY), stack.size());

    // This should fail
    try testing.expectError(Stack.Error.StackOverflow, stack.append(1000));
}

test "Stack: data alignment and access" {
    var stack = try Stack.init(testing.allocator);
    defer stack.deinit(std.testing.allocator);

    // Test that we can store large values
    const large_value: u256 = 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF;
    try stack.append(large_value);

    try testing.expectEqual(large_value, try stack.pop());
}

File: src/guillotine-go/stack/stack_test.go
package stack

import (
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"

	"github.com/evmts/guillotine/bindings/go/primitives"
)

// Test the stack creation and destruction
func TestStack_New(t *testing.T) {
	t.Run("ValidCreation", func(t *testing.T) {
		// Should create a new stack successfully
		stack, err := New()
		require.NoError(t, err)
		require.NotNil(t, stack)
		defer stack.Close()

		// Should start empty
		assert.True(t, stack.IsEmpty())
		assert.Equal(t, uint32(0), stack.Size())
		assert.False(t, stack.IsFull())
		assert.Equal(t, uint32(1024), stack.Capacity()) // EVM standard
	})
}

func TestStack_BasicOperations(t *testing.T) {
	stack, err := New()
	require.NoError(t, err)
	defer stack.Close()

	t.Run("PushU64", func(t *testing.T) {
		// Should push 64-bit values
		err := stack.PushU64(42)
		require.NoError(t, err)
		assert.Equal(t, uint32(1), stack.Size())
		assert.False(t, stack.IsEmpty())

		err = stack.PushU64(100)
		require.NoError(t, err)
		assert.Equal(t, uint32(2), stack.Size())
	})

	t.Run("PopU64", func(t *testing.T) {
		// Should pop in LIFO order
		value, err := stack.PopU64()
		require.NoError(t, err)
		assert.Equal(t, uint64(100), value)
		assert.Equal(t, uint32(1), stack.Size())

		value, err = stack.PopU64()
		require.NoError(t, err)
		assert.Equal(t, uint64(42), value)
		assert.Equal(t, uint32(0), stack.Size())
		assert.True(t, stack.IsEmpty())
	})

	t.Run("PopEmptyStack", func(t *testing.T) {
		// Should error on underflow
		_, err := stack.PopU64()
		assert.Error(t, err)
		assert.Contains(t, err.Error(), "underflow")
	})
}

func TestStack_U256Operations(t *testing.T) {
	stack, err := New()
	require.NoError(t, err)
	defer stack.Close()

	t.Run("PushU256", func(t *testing.T) {
		// Test with various U256 values
		values := []string{
			"0x0",
			"0x42",
			"0x1234567890abcdef",
			"0x1fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff",
		}

		for _, valueHex := range values {
			value, err := primitives.U256FromHex(valueHex)
			require.NoError(t, err)

			err = stack.PushU256(value)
			require.NoError(t, err)
		}

		assert.Equal(t, uint32(len(values)), stack.Size())
	})

	t.Run("PopU256", func(t *testing.T) {
		// Should pop in reverse order
		expectedValues := []string{
			"0x1fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff",
			"0x1234567890abcdef", 
			"0x42",
			"0x0",
		}

		for _, expectedHex := range expectedValues {
			expected, err := primitives.U256FromHex(expectedHex)
			require.NoError(t, err)

			actual, err := stack.PopU256()
			require.NoError(t, err)
			assert.True(t, actual.Equal(expected), 
				"Expected %s, got %s", expected.Hex(), actual.Hex())
		}

		assert.True(t, stack.IsEmpty())
	})
}

func TestStack_PeekOperations(t *testing.T) {
	stack, err := New()
	require.NoError(t, err)
	defer stack.Close()

	// Set up test data
	values := []uint64{10, 20, 30, 40, 50}
	for _, v := range values {
		err := stack.PushU64(v)
		require.NoError(t, err)
	}

	t.Run("PeekTop", func(t *testing.T) {
		// Should peek at top without removing
		value, err := stack.PeekU64()
		require.NoError(t, err)
		assert.Equal(t, uint64(50), value) // Last pushed
		assert.Equal(t, uint32(5), stack.Size()) // Size unchanged
	})

	t.Run("PeekAt", func(t *testing.T) {
		// Should peek at specific depths
		testCases := []struct {
			depth    uint32
			expected uint64
		}{
			{0, 50}, // Top
			{1, 40}, // Second from top
			{2, 30}, // Third from top  
			{3, 20}, // Fourth from top
			{4, 10}, // Bottom
		}

		for _, tc := range testCases {
			value, err := stack.PeekAt(tc.depth)
			require.NoError(t, err, "Failed to peek at depth %d", tc.depth)
			assert.Equal(t, tc.expected, value, 
				"Wrong value at depth %d", tc.depth)
		}

		// Size should remain unchanged
		assert.Equal(t, uint32(5), stack.Size())
	})

	t.Run("PeekOutOfBounds", func(t *testing.T) {
		// Should error when peeking beyond stack size
		_, err := stack.PeekAt(10)
		assert.Error(t, err)
		assert.Contains(t, err.Error(), "index")
	})
}

func TestStack_EVMOperations(t *testing.T) {
	stack, err := New()
	require.NoError(t, err)
	defer stack.Close()

	// Set up test data: [10, 20, 30, 40, 50] (50 on top)
	for _, v := range []uint64{10, 20, 30, 40, 50} {
		err := stack.PushU64(v)
		require.NoError(t, err)
	}

	t.Run("Dup", func(t *testing.T) {
		// DUP1 - duplicate top element
		err := stack.Dup(1)
		require.NoError(t, err)
		assert.Equal(t, uint32(6), stack.Size())

		// Top two should be same
		val1, err := stack.PopU64()
		require.NoError(t, err)
		val2, err := stack.PopU64()
		require.NoError(t, err)
		assert.Equal(t, uint64(50), val1)
		assert.Equal(t, uint64(50), val2)
	})

	t.Run("Swap", func(t *testing.T) {
		// Current stack: [10, 20, 30, 40] (40 on top)
		// SWAP1 - swap top two elements
		err := stack.Swap(1)
		require.NoError(t, err)
		assert.Equal(t, uint32(4), stack.Size()) // Size unchanged

		// Should have swapped 40 and 30
		val1, err := stack.PopU64()
		require.NoError(t, err)
		val2, err := stack.PopU64()  
		require.NoError(t, err)
		assert.Equal(t, uint64(30), val1) // Was second, now top
		assert.Equal(t, uint64(40), val2) // Was top, now second
	})
}

func TestStack_OverflowHandling(t *testing.T) {
	stack, err := New()
	require.NoError(t, err)
	defer stack.Close()

	t.Run("StackOverflow", func(t *testing.T) {
		// Fill stack to capacity (1024 items)
		capacity := stack.Capacity()
		for i := uint32(0); i < capacity; i++ {
			err := stack.PushU64(uint64(i))
			require.NoError(t, err)
		}

		assert.True(t, stack.IsFull())
		assert.Equal(t, capacity, stack.Size())

		// One more push should fail
		err := stack.PushU64(9999)
		assert.Error(t, err)
		assert.Contains(t, err.Error(), "overflow")
		
		// Size should be unchanged
		assert.Equal(t, capacity, stack.Size())
	})
}

func TestStack_ResourceManagement(t *testing.T) {
	t.Run("Close", func(t *testing.T) {
		stack, err := New()
		require.NoError(t, err)

		// Add some data
		err = stack.PushU64(42)
		require.NoError(t, err)

		// Should close successfully
		err = stack.Close()
		assert.NoError(t, err)

		// Operations after close should fail
		err = stack.PushU64(100)
		assert.Error(t, err)
		assert.Contains(t, err.Error(), "closed")
	})

	t.Run("DoubleClose", func(t *testing.T) {
		stack, err := New()
		require.NoError(t, err)

		err1 := stack.Close()
		err2 := stack.Close()
		
		assert.NoError(t, err1)
		assert.NoError(t, err2) // Should not error on double close
	})
}

func TestStack_Reset(t *testing.T) {
	stack, err := New()
	require.NoError(t, err)
	defer stack.Close()

	// Add some data
	for i := 0; i < 10; i++ {
		err := stack.PushU64(uint64(i))
		require.NoError(t, err)
	}
	assert.Equal(t, uint32(10), stack.Size())

	// Reset should clear the stack
	err = stack.Reset()
	require.NoError(t, err)

	assert.True(t, stack.IsEmpty())
	assert.Equal(t, uint32(0), stack.Size())
	assert.False(t, stack.IsFull())
}

func TestStack_Contents(t *testing.T) {
	stack, err := New()
	require.NoError(t, err)
	defer stack.Close()

	// Add test data
	testValues := []uint64{10, 20, 30, 40, 50}
	for _, v := range testValues {
		err := stack.PushU64(v)
		require.NoError(t, err)
	}

	t.Run("GetContents", func(t *testing.T) {
		contents, err := stack.GetContents()
		require.NoError(t, err)
		
		// Should get all items (top first)
		assert.Len(t, contents, len(testValues))
		
		// Verify order (top to bottom)
		expectedOrder := []uint64{50, 40, 30, 20, 10}
		for i, expected := range expectedOrder {
			assert.True(t, contents[i].Equal(primitives.NewU256(expected)),
				"Position %d: expected %d, got %s", i, expected, contents[i].Hex())
		}
	})
}

func TestStack_ConcurrentAccess(t *testing.T) {
	stack, err := New()
	require.NoError(t, err)
	defer stack.Close()

	t.Run("ConcurrentReads", func(t *testing.T) {
		// Add some data
		for i := 0; i < 100; i++ {
			err := stack.PushU64(uint64(i))
			require.NoError(t, err)
		}

		// Concurrent reads should be safe
		done := make(chan bool, 2)

		go func() {
			for i := 0; i < 50; i++ {
				size := stack.Size()
				assert.GreaterOrEqual(t, size, uint32(50)) // Should have at least 50
			}
			done <- true
		}()

		go func() {
			for i := 0; i < 50; i++ {
				isEmpty := stack.IsEmpty()
				assert.False(t, isEmpty) // Should not be empty
			}
			done <- true
		}()

		<-done
		<-done
	})
}

// Benchmark tests
func BenchmarkStack_PushU64(b *testing.B) {
	stack, err := New()
	if err != nil {
		b.Fatal(err)
	}
	defer stack.Close()

	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		if err := stack.PushU64(uint64(i)); err != nil {
			b.Fatal(err)
		}
		
		// Reset when approaching capacity
		if i%1000 == 999 {
			stack.Reset()
		}
	}
}

func BenchmarkStack_PopU64(b *testing.B) {
	stack, err := New()
	if err != nil {
		b.Fatal(err)
	}
	defer stack.Close()

	// Pre-populate stack
	for i := 0; i < 1000; i++ {
		if err := stack.PushU64(uint64(i)); err != nil {
			b.Fatal(err)
		}
	}

	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		// Replenish when empty
		if stack.IsEmpty() {
			for j := 0; j < 1000; j++ {
				stack.PushU64(uint64(j))
			}
		}
		
		if _, err := stack.PopU64(); err != nil {
			b.Fatal(err)
		}
	}
}